CORE-8622 Recreate transactions with empty component groups

lankydan opened the pull request:
Empty component groups are not recorded when a transaction is persisted to the database.

When loading a transaction from the database it finds all the component groups with the matching transaction id. These component groups are then used to recreate a `WireTransaction`.

The issue is that the empty component groups which were not found in the database, because we never recorded their existence has an effect on the calculated transaction id in `WireTransaction`. Causing the loaded transaction to be subtly different and also have a different transaction id than it was saved with.

To fix this, the component group indexes of the transaction are recorded in the `TransactionMetadata` when initially creating a `WireTransaction`. The metadata is used when recreating `WireTransaction`s from the database, where we now know which groups should exist and can therefore create an empty component group if the database did not return anything for a group that should exist.

vlajos commented:
Hmm. This current PR looks overlapping with this:
https://github.com/corda/corda-runtime-os/pull/2563
Could we merge that one first and then simplify this PR?

lankydan committed:
CORE-8622 Recreate transactions with empty component groups

Empty component groups are not recorded when a transaction is persisted
to the database.

When loading a transaction from the database it finds all the component
groups with the matching transaction id. These component groups are then
used to recreate a `WireTransaction`.

The issue is that the empty component groups which were not found in the
database, because we never recorded their existence has an effect on the
calculated transaction id in `WireTransaction`. Causing the loaded
transaction to be subtly different and also have a different transaction
id than it was saved with.

To fix this, the component group indexes of the transaction are recorded
in the `TransactionMetadata` when initially creating a
`WireTransaction`. The metadata is used when recreating
`WireTransaction`s from the database, where we now know which groups
should exist and can therefore create an empty component group if the
database did not return anything for a group that should exist.

vlajos commented on testing/ledger/ledger-common-testkit/src/main/kotlin/net/corda/ledger/common/testkit/TransactionMetadataExample.kt:
```diff
@@ -0,0 +1,22 @@
+            TransactionMetadata.CPI_METADATA_KEY to cpiMetadata,
+            TransactionMetadata.CPK_METADATA_KEY to cpkMetadata,
+            TransactionMetadata.SCHEMA_VERSION_KEY to TransactionMetadata.SCHEMA_VERSION
+        )
```
If/When the json validation starts working, it may start to miss the new componentGroupIndexes section.
(not a blocker though.)

lankydan commented:
API PR https://github.com/corda/corda-api/pull/718

mnesbit commented on libs/ledger/ledger-common-data/src/main/resources/schema/transaction-metadata.json:
```diff
@@ -126,6 +126,14 @@
+    "componentGroupIndexes": {
+      "title": "Component group indexes",
+      "description": "The component groups contained in the transaction, to its 0-based index within the top-level Merkle tree",
+      "type": "object",
```
I was hoping we would end up with a JSON structure such as :
```
"componentGroups" : [
     [
         "metadata"
     ],
     "net.corda.v5.ledger.StateRef",
     "net.corda.v5.ledger.Command",
     [
        "PublicKey",
         "TimeWindow"     
     ]
]
```

The offset in the array is the component group index. The entries are either class name if uniform across all leaves of component group, or an array of individual leaf types e.g. for Notary Info, or metadata component groups

lankydan replied:
Fair enough. There is this ticket that Oisin started, https://r3-cev.atlassian.net/browse/CORE-7269, which implemented something along the lines of what you're suggesting.

I'll link this comment in the jira and address it there separately.

blsemo reviewed (dismissed):
LGTM

lankydan committed:
CORE-8622 Fix tests

vlajos commented on libs/ledger/ledger-common-data/src/main/kotlin/net/corda/ledger/common/data/transaction/factory/WireTransactionFactoryImpl.kt:
```diff
@@ -42,39 +41,55 @@ class WireTransactionFactoryImpl @Activate constructor(
+        val completeComponentGroupLists = metadata
+            .getComponentGroupIndexes()
+            .map { index -> componentGroupLists.getOrElse(index) { arrayListOf() } }
+
```
Does this cover the case when we add new component groups after the transaction has already been created?

lankydan replied:
Yes.

Because the metadata holds the number of component groups, then if the component group enums change after the transaction is created the transaction is unaffected as it pulls the groups from its metadata. Furthermore, similar to how the transaction id was changing for missing groups, we do not want to add extra empty lists to a transaction when loading it from the database as that will also change the transaction id merkle proof.

lankydan committed:
CORE-7854 use number of component groups

lankydan committed:
CORE-7854 use number of component groups

lankydan committed:
CORE-7854 version

lankydan committed:
CORE-8622 fix

vlajos reviewed (approved):
LGTM

ronanbrowne reviewed (approved):
LGTM (build)
